casper/src/main/scala/coop/rchain/casper/util/DoublyLinkedDagOperations.scala

```diff
@@ -0,0 +1,100 @@
+package coop.rchain.casper.util
+
+import coop.rchain.casper.Estimator.BlockHash
+
+import scala.collection.immutable.{HashMap, HashSet}
+
+trait DoublyLinkedDag[A] {
+  val parentToChildAdjacencyList: Map[A, Set[A]]
+  val childToParentAdjacencyList: Map[A, Set[A]]
+  val dependencyFree: Set[A]
+}
+case class BlockDependencyDag(parentToChildAdjacencyList: Map[BlockHash, Set[BlockHash]],
+                              childToParentAdjacencyList: Map[BlockHash, Set[BlockHash]],
+                              dependencyFree: Set[BlockHash])
+    extends DoublyLinkedDag[BlockHash]
+
+object BlockDependencyDag {
+  def empty: BlockDependencyDag =
+    new BlockDependencyDag(
+      HashMap.empty[BlockHash, Set[BlockHash]],
+      HashMap.empty[BlockHash, Set[BlockHash]],
+      HashSet.empty[BlockHash]
+    )
+}
+
+object MapHelper {
+  def updatedWith[A, B](map: Map[A, B], key: A)(default: B)(f: B => B): Map[A, B] = {
+    val newValue = map.get(key).fold(default)(f)
+    map.updated(key, newValue)
+  }
+}
+
+object DoublyLinkedDagOperations {
+  def add[A](dag: DoublyLinkedDag[A], parent: A, child: A): DoublyLinkedDag[A] = {
+    val parentToChildAdjacencyList: Map[A, Set[A]] = dag.parentToChildAdjacencyList
+    val childToParentAdjacencyList: Map[A, Set[A]] = dag.childToParentAdjacencyList
+    val dependencyFree: Set[A]                     = dag.dependencyFree
+
+    assert(!parentToChildAdjacencyList.values.toSet.contains(Set.empty[A]))
```

Comment:
Wouldn't it be faster to do the one non-empty check on L47 rather than these full data-structure non-empty asserts?

Reply:
Please remove any extraneous asserts